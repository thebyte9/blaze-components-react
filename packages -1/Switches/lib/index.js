"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var utils_1 = __importDefault(require("@blaze-react/utils"));
var react_1 = __importStar(require("react"));
var Switches = utils_1["default"](function (_a) {
    var _b;
    var labelPosition = _a.labelPosition, onChange = _a.onChange, options = _a.options, modifier = _a.modifier, returnBoolean = _a.returnBoolean, error = _a.error, validationMessage = _a.validationMessage, _c = _a.utils, uniqueId = _c.uniqueId, classNames = _c.classNames, ErrorMessage = _c.ErrorMessage, attrs = __rest(_a, ["labelPosition", "onChange", "options", "modifier", "returnBoolean", "error", "validationMessage", "utils"]);
    var _d = Array.isArray(options)
        ? {
            formatedOptions: options,
            wrap: function (child) { return (react_1["default"].createElement("div", { className: "form-group form-group--switch" }, child)); }
        }
        : {
            formatedOptions: [options],
            wrap: function (child) { return react_1["default"].createElement(react_1["default"].Fragment, null, child); }
        }, wrap = _d.wrap, formatedOptions = _d.formatedOptions;
    var _e = react_1.useState(formatedOptions), data = _e[0], setData = _e[1];
    var toggle = function (_a) {
        var event = _a.event, item = _a.item, key = _a.key;
        if (item.disabled) {
            return;
        }
        data[key].checked = !item.checked;
        setData(data.slice());
        var checked = data.filter(function (option) { return option.checked; });
        if (returnBoolean) {
            onChange({ event: event, value: !!checked.length, data: data });
            return;
        }
        onChange({ event: event, value: checked, data: data });
    };
    var switchClassNames = classNames("switch", (_b = {},
        _b["switch--" + modifier] = !!modifier,
        _b["switch--label--" + labelPosition] = !!labelPosition,
        _b));
    return (react_1["default"].createElement(react_1.Fragment, null,
        wrap(data.map(function (item, key) {
            var _a = item.checked, checked = _a === void 0 ? false : _a, value = item.value, disabled = item.disabled, required = item.required, label = item.label, _b = item.id, id = _b === void 0 ? uniqueId(item) : _b;
            return (react_1["default"].createElement("div", { className: switchClassNames, key: id },
                react_1["default"].createElement("div", { className: "switch__text" }, label),
                react_1["default"].createElement("div", { className: "switch__item" },
                    react_1["default"].createElement("input", __assign({ readOnly: true, type: "checkbox", value: value, disabled: disabled, checked: checked, required: required, onChange: function (event) { return toggle({ event: event, item: item, key: key }); }, id: id }, attrs)),
                    react_1["default"].createElement("label", { htmlFor: id }, "toggle"))));
        })),
        error && react_1["default"].createElement(ErrorMessage, { message: validationMessage })));
});
var availableModifiers = {
    checked: "checked",
    "default": "default",
    disabled: "disabled",
    primary: "primary",
    secondary: "secondary",
    unchecked: "unchecked"
};
var availablePositions = {
    base: "base",
    left: "left",
    right: "right",
    top: "top"
};
Switches.availableModifiers = availableModifiers;
Switches.availablePositions = availablePositions;
Switches.defaultProps = {
    error: false,
    labelPosition: "right",
    modifier: "",
    returnBoolean: false,
    validationMessage: "This field is required"
};
exports["default"] = Switches;
