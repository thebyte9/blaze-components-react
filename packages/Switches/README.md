## Description

Switches component is a digital on/off switch used for activating one of two predefined options.

## Usage

- Group

```js
const multiple = [
  {
    id: 1,
    label: "Switch text"
  },
  {
    id: 2,
    label: "Switch text"
  },
  {
    disabled: true,
    id: 4,
    label: "Disabled"
  }
];

<Switches options={multiple} onChange={({ event, checked }) => {}} />;
```

- Single

```js
const single = {
  label: "Switch text",
  required: true
};

<Switches
  options={single}
  returnBoolean
  onChange={({ event, checked }) => {}}
```

## API

##### Switches can receive a number of `props` as follow:

| NAME          |      TYPE       | DEFAULT  | OPTIONS                                                           |
| :------------ | :-------------: | :------: | ----------------------------------------------------------------- |
| modifier      |     String      |  empty   | primary, secondary, default, unchecked disabled, checked disabled |
| labelPosition |     String      |  right   | right, left, top, base                                            |
| options       | Array or Object |    []    |
| disabled      |     Boolean     |  false   |
| required      |     Boolean     |  false   |
| returnBoolean |     Boolean     |  false   |
| onChange      |    Function     | () => {} |
