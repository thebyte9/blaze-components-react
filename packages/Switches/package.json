{
  "name": "@blaze-react/switches",
  "version": "0.4.15",
  "publishConfig": {
    "access": "public"
  },
  "description": "",
  "main": "./lib/index.js",
  "scripts": {
    "build": "../../node_modules/.bin/tsc --jsx react --esModuleInterop ./src/* --outDir lib ",
    "build-storybook": "build-storybook -c .storybook -o docs-build",
    "lint": "../../node_modules/.bin/tslint -c ../../tslint.json src/**/*.tsx",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "pre-publish": "npm run build && npm run lint && npm run test"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@blaze-react/utils": "^0.4.15"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.4",
    "@babel/core": "^7.4.5",
    "@blaze-react/blaze-components-theme": "^0.4.14",
    "@types/uuid": "^3.4.4"
  },
  "peerDependencies": {
    "prop-types": "^15.7.2",
    "react": "^16.9.0",
    "react-dom": "^16.9.0"
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ]
  },
  "gitHead": "c7cc1d26b4046b800c727feb1dd283dbd833328c"
}
