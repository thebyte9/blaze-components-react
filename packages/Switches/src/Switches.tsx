import React, { Fragment, useState } from 'react';

import { ErrorMessage } from '@blaze-react/utils';
import { buildClassNames } from '@blaze-react/utils';

interface IOptions {
  checked: boolean;
  value: string;
  disabled: boolean;
  required: boolean;
  label: string;
  id?: string;
  tooltip?: string | JSX.Element;
}

type TlabelPosition = 'right' | 'left' | 'base' | 'top';

type TModifiers = 'checked' | 'default' | 'disabled' | 'primary' | 'secondary' | 'unchecked';

interface ISwitchesProps {
  labelPosition?: TlabelPosition;
  options: IOptions[] | IOptions;
  modifier?: TModifiers;
  returnBoolean?: boolean;
  error?: boolean;
  validationMessage: string | JSX.Element;
  onChange: ({
    event,
    value,
    data,
  }: {
    event: React.ChangeEvent<HTMLInputElement>;
    value: IOptions[] | boolean;
    data: IOptions[];
  }) => void;
}

const Switches = ({
  labelPosition,
  onChange,
  options,
  modifier,
  returnBoolean,
  error,
  validationMessage,
  tooltip,
  ...attrs
}: ISwitchesProps): JSX.Element => {
  const {
    wrap,
    formatedOptions,
  }: {
    wrap: (child: JSX.Element[]) => JSX.Element;
    formatedOptions: IOptions[];
  } = Array.isArray(options)
      ? {
        formatedOptions: options,
        wrap: (child: JSX.Element[]): JSX.Element => <div className="form-group form-group--switch">{child}</div>,
      }
      : {
        formatedOptions: [options],
        wrap: (child: JSX.Element[]): JSX.Element => <>{child}</>,
      };

  const [data, setData] = useState<IOptions[]>(formatedOptions);

  const toggle = ({
    event,
    item,
    key,
  }: {
    event: React.ChangeEvent<HTMLInputElement>;
    item: IOptions;
    key: number;
  }): void => {
    if (item.disabled) {
      return;
    }

    data[key].checked = !item.checked;
    setData([...data]);

    const checked: IOptions[] = data.filter((option: IOptions): boolean => option.checked);

    if (returnBoolean) {
      onChange({ event, value: !!checked.length, data });
      return;
    }

    onChange({ event, value: checked, data });
  };

  const switchClassNames: string = buildClassNames('switch', {
    [`switch--${modifier}`]: !!modifier,
    [`switch--label--${labelPosition}`]: !!labelPosition,
  });

  return (
    <Fragment>
      {wrap(
        data.map((item: IOptions, key: number): JSX.Element => {
          const { checked = false, value, disabled, required, label, id = `switch-${key}` } = item;

          const labelWithTooltip = (
            <>
              {label}
              {tooltip}
            </>
          );

          return (
            <div className={switchClassNames} key={id}>
              <div className="switch__text">{labelWithTooltip}</div>
              <div className="switch__item">
                <input
                  readOnly
                  type="checkbox"
                  value={value}
                  disabled={disabled}
                  checked={checked}
                  required={required}
                  onChange={(event: React.ChangeEvent<HTMLInputElement>): void => toggle({ event, item, key })}
                  id={id}
                  {...attrs}
                />
                <label htmlFor={id}>toggle</label>
              </div>
            </div>
          );
        }),
      )}
      {error && <ErrorMessage message={validationMessage} />}
    </Fragment>
  );
};

const availableModifiers = {
  checked: 'checked',
  default: 'default',
  disabled: 'disabled',
  primary: 'primary',
  secondary: 'secondary',
  unchecked: 'unchecked',
};

const availablePositions = {
  base: 'base',
  left: 'left',
  right: 'right',
  top: 'top',
};

Switches.availableModifiers = availableModifiers;
Switches.availablePositions = availablePositions;

Switches.defaultProps = {
  error: false,
  labelPosition: 'right',
  modifier: '',
  returnBoolean: false,
  validationMessage: 'This field is required',
};

export default Switches;
