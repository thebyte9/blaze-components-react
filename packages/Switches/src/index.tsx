import withUtils from "@blaze-react/utils";
import React, { Fragment, FunctionComponent, useState } from "react";

interface IOptions {
  checked: boolean;
  value: string;
  disabled: boolean;
  required: boolean;
  label: string;
  id?: string;
}
interface IErrorMessage {
  message: string | JSX.Element;
  icon?: string;
}
type TlabelPosition = "right" | "left" | "base" | "top";
type TModifiers =
  | "checked"
  | "default"
  | "disabled"
  | "primary"
  | "secondary"
  | "unchecked";
interface ISwitchesProps {
  labelPosition?: TlabelPosition;
  options: IOptions[] | IOptions;
  modifier?: TModifiers;
  returnBoolean?: boolean;
  error?: boolean;
  validationMessage: string | JSX.Element;
  onChange: ({
    event,
    value,
    data,
  }: {
    event: React.ChangeEvent<HTMLInputElement>;
    value: IOptions[] | boolean;
    data: IOptions[];
  }) => void;
  utils: {
    uniqueId: (element: any) => string;
    classNames: (className: string | object, classNames?: object) => string;
    ErrorMessage: FunctionComponent<IErrorMessage>;
  };
}
const Switches = withUtils(
  ({
    labelPosition,
    onChange,
    options,
    modifier,
    returnBoolean,
    error,
    validationMessage,
    utils: { uniqueId, classNames, ErrorMessage },
    ...attrs
  }: ISwitchesProps): JSX.Element => {
    const {
      wrap,
      formatedOptions,
    }: {
      wrap: (child: JSX.Element[]) => JSX.Element;
      formatedOptions: IOptions[];
    } = Array.isArray(options)
      ? {
          formatedOptions: options,
          wrap: (child: JSX.Element[]): JSX.Element => (
            <div className="form-group form-group--switch">{child}</div>
          ),
        }
      : {
          formatedOptions: [options],
          wrap: (child: JSX.Element[]): JSX.Element => <>{child}</>,
        };

    const [data, setData] = useState<IOptions[]>(formatedOptions);

    const toggle = ({
      event,
      item,
      key,
    }: {
      event: React.ChangeEvent<HTMLInputElement>;
      item: IOptions;
      key: number;
    }): void => {
      if (item.disabled) {
        return;
      }

      data[key].checked = !item.checked;
      setData([...data]);

      const checked: IOptions[] = data.filter(
        (option: IOptions): boolean => option.checked
      );

      if (returnBoolean) {
        onChange({ event, value: !!checked.length, data });
        return;
      }

      onChange({ event, value: checked, data });
    };

    const switchClassNames: string = classNames("switch", {
      [`switch--${modifier}`]: !!modifier,
      [`switch--label--${labelPosition}`]: !!labelPosition,
    });

    return (
      <Fragment>
        {wrap(
          data.map(
            (item: IOptions, key: number): JSX.Element => {
              const {
                checked = false,
                value,
                disabled,
                required,
                label,
                id = uniqueId(item),
              } = item;

              return (
                <div className={switchClassNames} key={id}>
                  <div className="switch__text">{label}</div>
                  <div className="switch__item">
                    <input
                      readOnly
                      type="checkbox"
                      value={value}
                      disabled={disabled}
                      checked={checked}
                      required={required}
                      onChange={(
                        event: React.ChangeEvent<HTMLInputElement>
                      ): void => toggle({ event, item, key })}
                      id={id}
                      {...attrs}
                    />
                    <label htmlFor={id}>toggle</label>
                  </div>
                </div>
              );
            }
          )
        )}
        {error && <ErrorMessage message={validationMessage} />}
      </Fragment>
    );
  }
);
const availableModifiers: object = {
  checked: "checked",
  default: "default",
  disabled: "disabled",
  primary: "primary",
  secondary: "secondary",
  unchecked: "unchecked",
};

const availablePositions: object = {
  base: "base",
  left: "left",
  right: "right",
  top: "top",
};

Switches.availableModifiers = availableModifiers;
Switches.availablePositions = availablePositions;

Switches.defaultProps = {
  error: false,
  labelPosition: "right",
  modifier: "",
  returnBoolean: false,
  validationMessage: "This field is required",
};
export default Switches;
