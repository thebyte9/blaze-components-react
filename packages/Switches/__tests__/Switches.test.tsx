import { mount, shallow } from "enzyme";
import expect from "expect";
import React from "react";
import Switches from "../src";

const options = [
  {
    id: 1,
    label: "Switch text"
  },
  {
    id: 2,
    label: "Switch text"
  },
  {
    disabled: true,
    label: "Disabled"
  }
];

const single = {
  label: "Switch text",
  required: true
};

const defaultProps = (override = {}) => ({
  onChange: () => ({}),
  options,
  ...override
});

describe("Switches component", () => {
  test("should be defined and renders correctly (snapshot)", () => {
    const wrapper = shallow(<Switches {...defaultProps()} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test("should toggle Switch on click", () => {
    const override = {
      modifier: Switches.availableModifiers.secondary,
      options: single,
      returnBoolean: true
    };

    const wrapper = mount(<Switches {...defaultProps(override)} />);

    wrapper
      .find("input")
      .at(0)
      .simulate("change");

    expect(
      wrapper
        .find("input")
        .at(0)
        .prop("checked")
    ).toBe(true);
  });

  test("can't interact when Switch is disabled", () => {
    const wrapper = mount(<Switches {...defaultProps()} />);
    wrapper
      .find("input")
      .at(1)
      .simulate("change");
    expect(
      wrapper
        .find("input")
        .at(1)
        .prop("checked")
    ).toBe(true);

    wrapper
      .find("input")
      .at(2)
      .simulate("change");
    expect(
      wrapper
        .find("input")
        .at(2)
        .prop("checked")
    ).toBe(false);
  });
});
