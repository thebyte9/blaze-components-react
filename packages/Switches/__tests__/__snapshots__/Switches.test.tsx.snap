// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switches component should be defined and renders correctly (snapshot) 1`] = `
<Component
  error={false}
  labelPosition="right"
  modifier=""
  onChange={[Function]}
  options={
    Array [
      Object {
        "id": 1,
        "label": "Switch text",
      },
      Object {
        "id": 2,
        "label": "Switch text",
      },
      Object {
        "disabled": true,
        "label": "Disabled",
      },
    ]
  }
  returnBoolean={false}
  utils={
    Object {
      "ErrorMessage": [Function],
      "classNames": [Function],
      "uniqueId": [Function],
    }
  }
  validationMessage="This field is required"
/>
`;
